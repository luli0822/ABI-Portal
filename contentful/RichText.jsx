import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { BLOCKS, MARKS, INLINES } from '@contentful/rich-text-types';
import * as Contentful from '@contentful/rich-text-react-renderer';
import ContentfulLink from './ContentfulLink';
import TextReplacement from './TextReplacement';
import siteDataContext from './siteDataContext';

function RichText(props) {
	const { lang, fieldOrFallback } = useContext(siteDataContext);
	const { document, rootClass, replacements } = props;

	const options = {
		renderMark: {
			[MARKS.BOLD]: text => <span className={`${rootClass}-b`}>{text}</span>,
			[MARKS.ITALIC]: text => <span className={`${rootClass}-i`}>{text}</span>,
		},
		renderNode: {
			[BLOCKS.PARAGRAPH]: (node, children) => <p className={`${rootClass}-p`}>{children}</p>,
			[BLOCKS.UL_LIST]: (node, children) => <ul className={`${rootClass}-ul`}>{children}</ul>,
			[BLOCKS.LIST_ITEM]: (node, children) => <li className={`${rootClass}-li`}>{children}</li>,
			[INLINES.EMBEDDED_ENTRY]: node => {
				const link = fieldOrFallback(node.data.target);
				const linkTitle = fieldOrFallback(link.fields.title, lang);
				return (
					<ContentfulLink linkEntry={link} className={`${rootClass}-link`}>
						<TextReplacement text={linkTitle} replacements={replacements} />
					</ContentfulLink>
				);
			},
		},
		renderText: children => <TextReplacement text={children} replacements={replacements} />,
	};

	return Contentful.documentToReactComponents(document, options);
}

RichText.propTypes = {
	document: PropTypes.shape({}).isRequired,
	rootClass: PropTypes.string,
	replacements: PropTypes.shape({}),
};

RichText.defaultProps = {
	rootClass: 'rich-text',
	replacements: {},
};

export default RichText;
