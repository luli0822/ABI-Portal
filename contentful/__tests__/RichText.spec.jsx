import React from 'react';
import { mount } from 'enzyme';
import RichText from '../RichText';

import {
	mockLang,
	mockPathMap,
	mockLinkedEntries,
	mockLinkedAssets,
	mockFieldOrFallback,
	mockRichText,
} from '../__mocks__/SiteDataContext';

// Mock React useContext
jest.mock('react', () => {
	const ActualReact = require.requireActual('react');
	return {
		...ActualReact,
		useContext: () => ({
			pathMap: mockPathMap,
			fieldOrFallback: mockFieldOrFallback,
			linkedEntries: mockLinkedEntries,
			linkedAssets: mockLinkedAssets,
			lang: mockLang,
		}),
	};
});

const mockProps = {
	replacements: {
		italicized: <span className="i">italicized</span>,
	},
};

describe('<RichText /> ', () => {
	it('renders correctly', () => {
		const component = mount(<RichText document={mockRichText['without-markings']} />);

		expect(component.debug()).toMatchSnapshot();
	});
	it('renders marks correctly with correct classes', () => {
		const component = mount(<RichText document={mockRichText['with-markings']} />);

		expect(component.debug()).toMatchSnapshot();
	});
	it('renders correctly with contentful link', () => {
		// eslint-disable-next-line react/jsx-props-no-spreading
		const component = mount(<RichText document={mockRichText['with-link']} {...mockProps} />);

		expect(component.debug()).toMatchSnapshot();
	});
	it('renders correctly with Text Replacement markup', () => {
		const component = mount(
			// eslint-disable-next-line react/jsx-props-no-spreading
			<RichText document={mockRichText['without-markings']} {...mockProps} />
		);

		expect(component.debug()).toMatchSnapshot();
	});
	// it('Check getClient() and fetchSiteData() will return the expected site Data from contentful ', async () => {
	// 	// TODO: test1: create snapshot that renders rich text properly. Import context mocks.
	// 	// TODO: test2: Check mounting Block paragraph richText (using bold/italic). Check the proper classes are attached to element
	// 	// TODO: test3: Check mounting INLINES.EMBEDDED_ENTRY (contentfulLink). Check the proper classes are attached to element
	// 	// TODO: test4: Check mounting list items. Check the proper classes are attached to element

	// 	expect(1).toBe(1);
	// });
});
