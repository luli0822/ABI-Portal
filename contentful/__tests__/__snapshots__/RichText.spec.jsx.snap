// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RichText />  renders correctly 1`] = `
"<RichText document={{...}} rootClass=\\"rich-text\\" replacements={{...}}>
  <p className=\\"rich-text-p\\">
    <TextReplacement text=\\"Text without markings italicized\\" replacements={{...}}>
      Text without markings italicized
    </TextReplacement>
  </p>
</RichText>"
`;

exports[`<RichText />  renders correctly with Text Replacement markup 1`] = `
"<RichText document={{...}} replacements={{...}} rootClass=\\"rich-text\\">
  <p className=\\"rich-text-p\\">
    <TextReplacement text=\\"Text without markings italicized\\" replacements={{...}}>
      Text without markings 
      <span className=\\"i\\">
        italicized
      </span>
    </TextReplacement>
  </p>
</RichText>"
`;

exports[`<RichText />  renders correctly with contentful link 1`] = `
"<RichText document={{...}} replacements={{...}} rootClass=\\"rich-text\\">
  <p className=\\"rich-text-p\\">
    <span className=\\"rich-text-b\\">
      <TextReplacement text=\\"We bring marketing sites to life\\" replacements={{...}}>
        We bring marketing sites to life
      </TextReplacement>
    </span>
    <TextReplacement text=\\" for \\" replacements={{...}}>
       for 
    </TextReplacement>
    <span className=\\"rich-text-i\\">
      <TextReplacement text=\\"brands\\" replacements={{...}}>
        brands
      </TextReplacement>
    </span>
    <TextReplacement text=\\" across the organization. Learn more \\" replacements={{...}}>
       across the organization. Learn more 
    </TextReplacement>
    <ContentfulLink linkEntry={{...}} className=\\"rich-text-link\\" targetLang={[undefined]}>
      <a href=\\"www.site.com\\" aria-label=\\"accessible text \\" className=\\"rich-text-link\\">
        <TextReplacement text=\\"mock link title italicized\\" replacements={{...}}>
          mock link title 
          <span className=\\"i\\">
            italicized
          </span>
        </TextReplacement>
      </a>
    </ContentfulLink>
    <TextReplacement text=\\"\\" replacements={{...}} />
  </p>
  <p className=\\"rich-text-p\\">
    <TextReplacement text=\\"\\" replacements={{...}} />
  </p>
  <p className=\\"rich-text-p\\">
    <TextReplacement text=\\"\\" replacements={{...}} />
  </p>
</RichText>"
`;

exports[`<RichText />  renders marks correctly with correct classes 1`] = `
"<RichText document={{...}} rootClass=\\"rich-text\\" replacements={{...}}>
  <p className=\\"rich-text-p\\">
    <span className=\\"rich-text-b\\">
      <TextReplacement text=\\"Bold Text\\" replacements={{...}}>
        Bold Text
      </TextReplacement>
    </span>
    <TextReplacement text=\\" for \\" replacements={{...}}>
       for 
    </TextReplacement>
    <span className=\\"rich-text-i\\">
      <TextReplacement text=\\"Italic Text\\" replacements={{...}}>
        Italic Text
      </TextReplacement>
    </span>
  </p>
  <ul className=\\"rich-text-ul\\">
    <li className=\\"rich-text-li\\">
      <p className=\\"rich-text-p\\">
        <TextReplacement text=\\"point 1\\" replacements={{...}}>
          point 1
        </TextReplacement>
      </p>
    </li>
    <li className=\\"rich-text-li\\">
      <p className=\\"rich-text-p\\">
        <TextReplacement text=\\"point 2\\" replacements={{...}}>
          point 2
        </TextReplacement>
      </p>
    </li>
  </ul>
</RichText>"
`;
